2.0: Replicating the conflicted document causes unexpected behavior (crash).

```
09-11 15:57:11.014 19396-19396/? E/AndroidRuntime: FATAL EXCEPTION: main
 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.couchbase.todo/com.couchbase.todo.ListsActivity}: java.lang.NullPointerException
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2180)
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2230)
     at android.app.ActivityThread.access$600(ActivityThread.java:141)
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1234)
     at android.os.Handler.dispatchMessage(Handler.java:99)
     at android.os.Looper.loop(Looper.java:137)
     at android.app.ActivityThread.main(ActivityThread.java:5039)
     at java.lang.reflect.Method.invokeNative(Native Method)
     at java.lang.reflect.Method.invoke(Method.java:511)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
     at dalvik.system.NativeStart.main(Native Method)
  Caused by: java.lang.NullPointerException
     at com.couchbase.lite.LiveQuery.run(LiveQuery.java:59)
     at com.couchbase.todo.LiveListsAdapter.<init>(LiveListsAdapter.java:53)
     at com.couchbase.todo.ListsActivity.onCreate(ListsActivity.java:68)
     at android.app.Activity.performCreate(Activity.java:5104)
     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1080)
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2144)
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2230) 
     at android.app.ActivityThread.access$600(ActivityThread.java:141) 
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1234) 
     at android.os.Handler.dispatchMessage(Handler.java:99) 
     at android.os.Looper.loop(Looper.java:137) 
     at android.app.ActivityThread.main(ActivityThread.java:5039) 
     at java.lang.reflect.Method.invokeNative(Native Method) 
     at java.lang.reflect.Method.invoke(Method.java:511) 
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560) 
     at dalvik.system.NativeStart.main(Native Method) 
```

--- a/shared/src/main/java/com/couchbase/lite/Database.java
+++ b/shared/src/main/java/com/couchbase/lite/Database.java
@@ -751,9 +751,9 @@ public final class Database implements C4Constants {
                     Log.i(TAG, "Resolving doc '%s' with %s (mine=%s, theirs=%s, base=%s)",
                             docID,
                             resolver != null ? resolver.getClass().getSimpleName() : "null",
-                            doc != null ? doc.getRevID() : "null",
-                            otherDoc != null ? otherDoc.getRevID() : "null",
-                            baseDoc != null ? baseDoc.getRevID() : "null");
+                            doc != null ? doc.getSelectedRevID() : "null",
+                            otherDoc != null ? otherDoc.getSelectedRevID() : "null",
+                            baseDoc != null ? baseDoc.getSelectedRevID() : "null");
                     resolved = resolver.resolve(conflict);
                     if (resolved == null)
                         throw new CouchbaseLiteException(LiteCoreDomain, kC4ErrorConflict);
@@ -764,11 +764,11 @@ public final class Database implements C4Constants {
                 String losingRevID;
                 byte[] mergedBody = null;
                 if (resolved == otherDoc) {
-                    winningRevID = otherDoc.getRevID();
-                    losingRevID = doc.getRevID();
+                    winningRevID = otherDoc.getSelectedRevID();
+                    losingRevID = doc.getSelectedRevID();
                 } else {
-                    winningRevID = doc.getRevID();
-                    losingRevID = otherDoc.getRevID();
+                    winningRevID = doc.getSelectedRevID();
+                    losingRevID = otherDoc.getSelectedRevID();
                     if (resolved != doc) {
                         resolved.setDatabase(this);
                         mergedBody = resolved.encode();

--- a/shared/src/main/java/com/couchbase/lite/LiveQuery.java
+++ b/shared/src/main/java/com/couchbase/lite/LiveQuery.java
@@ -54,6 +54,11 @@ public class LiveQuery implements DatabaseChangeListener {
      * Starts observing database changes and reports changes in the query result.
      */
     public void run() {
+        if (query == null)
+            throw new IllegalArgumentException("query should not be null.");
+        if (query.getDatabase() == null)
+            throw new IllegalArgumentException("associated database should not be null.");
+
         observing = true;
         releaseResultSet();
         query.getDatabase().addChangeListener(this);

--- a/shared/src/main/java/com/couchbase/lite/ReadOnlyDocument.java
+++ b/shared/src/main/java/com/couchbase/lite/ReadOnlyDocument.java
@@ -82,7 +82,7 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
      * @return true if deleted, false otherwise
      */
     public boolean isDeleted() {
-        return c4doc != null ? c4doc.getRawDoc().deleted() : false;
+        return c4doc != null && c4doc.getRawDoc() != null ? c4doc.getRawDoc().deleted() : false;
     }
 
     @Override
@@ -95,8 +95,8 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
     //---------------------------------------------
 
     // Sets c4doc and updates my root dictionary
-    protected void setC4Doc(CBLC4Doc c4doc) {
-        free();
+    protected synchronized void setC4Doc(CBLC4Doc c4doc) {
+        // NOTE: don't call free(), both c4doc and this.c4doc point to same C4Document.
 
         this.c4doc = c4doc;
         if (c4doc != null) {
@@ -119,9 +119,9 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
     //---------------------------------------------
 
     void free() {
-        if (c4doc != null) {
-            c4doc.free();
-            c4doc = null;
+        if (this.c4doc != null) {
+            this.c4doc.free();
+            this.c4doc = null;
         }
     }
 
@@ -145,7 +145,7 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
     // Document overrides this
     long generation() {
         // TODO: c4rev_getGeneration
-        return generationFromRevID(getRevID());
+        return generationFromRevID(getSelectedRevID());
     }
 
     /**
@@ -172,6 +172,7 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
 
     // Reads the document from the db into a new C4Document and returns it, w/o affecting my state.
     C4Document readC4Doc(boolean mustExist) throws LiteCoreException {
+        if (database == null || database.getC4Database() == null) throw new IllegalStateException();
         return database.getC4Database().get(getId(), mustExist);
     }
 
@@ -186,25 +187,25 @@ public class ReadOnlyDocument extends ReadOnlyDictionary {
             c4doc.getRawDoc().selectNextLeafRevision(false, true);
             setC4Doc(c4doc); // self.c4Doc = _c4Doc; // This will update to the selected revision
         } catch (LiteCoreException e) {
-            Log.e(TAG, "Failed to selectNextLeaf: doc -> " + c4doc, e);
+            Log.e(TAG, "Failed to selectNextLeaf: doc -> %s", e, c4doc);
             return false;
         }
         return true;
     }
 
     boolean selectCommonAncestor(ReadOnlyDocument doc1, ReadOnlyDocument doc2) {
-        if (!c4doc.getRawDoc().selectCommonAncestorRevision(doc1.getRevID(), doc2.getRevID()))
+        if (!c4doc.getRawDoc().selectCommonAncestorRevision(doc1.getSelectedRevID(), doc2.getSelectedRevID()))
             return false;
         setC4Doc(c4doc); // self.c4Doc = _c4Doc; // This will update to the selected revision
         return true;
     }
 
-    String getRevID() {
+    String getSelectedRevID() {
         return c4doc != null ? c4doc.getSelectedRevID() : null;
     }
 
     // Document overrides this
     byte[] encode() {
-        return c4doc.getRawDoc().getSelectedBody();
+        return c4doc != null ? c4doc.getRawDoc().getSelectedBody() : null;
     }
 }

